lnst/Recipes/ENRT/LinuxBridgeRecipe.py

```diff
@@ -0,0 +1,116 @@
+from lnst.Common.Parameters import IPv4NetworkParam, IPv6NetworkParam
+from lnst.Common.IpAddress import interface_addresses
+from lnst.Controller import HostReq, DeviceReq, RecipeParam
+from lnst.Recipes.ENRT.BaremetalEnrtRecipe import BaremetalEnrtRecipe
+from lnst.RecipeCommon.Ping.PingEndpoints import PingEndpoints
+from lnst.Recipes.ENRT.ConfigMixins.MTUHWConfigMixin import MTUHWConfigMixin
+from lnst.Devices import BridgeDevice
+
+
+class LinuxBridgeRecipe(MTUHWConfigMixin, BaremetalEnrtRecipe):
+    """
+    This recipe implements Enrt testing for a simple network scenario that looks
+    as follows
+
+    .. code-block:: none
+
+                    +--------+
+             +------+ switch +-----+
+             |      +--------+     |
+          +--+-+                 +-+--+
+        +-|eth0|-+             +-|eth0|-+
+        | +----+ |             | +----+ |
+        |   |    |             |   |    |
+        |  br0   |             |  br0   |
+        |        |             |        |
+        | host1  |             | host2  |
+        +--------+             +--------+
+
+    All sub configurations are included via Mixin classes.
+
+    The actual test machinery is implemented in the :any:`BaseEnrtRecipe` class.
+    """
+
+    host1 = HostReq()
+    host1.eth0 = DeviceReq(label="net1", driver=RecipeParam("driver"))
+
+    host2 = HostReq()
+    host2.eth0 = DeviceReq(label="net1", driver=RecipeParam("driver"))
+
+    net_ipv4 = IPv4NetworkParam(default="192.168.101.0/24")
+    net_ipv6 = IPv6NetworkParam(default="fc00::/64")
+
+    def test_wide_configuration(self):
+        """
+        Test wide configuration for this recipe involves adding the matched
+        NICs into a Linux bridge and configuring an IPv4 and IPv6 address
+        on the bridge device on both hosts.
+
+        host1.br0 = 192.168.101.1/24 and fc00::1/64
+
+        host2.br0 = 192.168.101.2/24 and fc00::2/64
+        """
+        host1, host2 = self.matched.host1, self.matched.host2
+        configuration = super().test_wide_configuration()
+        configuration.test_wide_devices = []
+
+        ipv4_addr = interface_addresses(self.params.net_ipv4)
+        ipv6_addr = interface_addresses(self.params.net_ipv6)
+
+        for i, host in enumerate([host1, host2]):
+            host.br0 = BridgeDevice()
+            host.eth0.down()
+            host.br0.slave_add(host.eth0)
+
+            for dev in [host.eth0, host.br0]:
```

Comment:
is a for loop really needed here, its the same amount of lines to do:

```
python
host.eth0.up()
host.br0.up()
```
```

Reply:
I think this should even be done automatically behind the scenes... Is it possible to get all created devices and call the `.up()` method on them in a single place? I don't think it's possible now, but it definitely could be. I'm just not sure if this is sane from a design perspective.

Reply:
> I think this should even be done automatically behind the scenes... Is it possible to get all created devices and call the `.up()` method on them in a single place? I don't think it's possible now, but it definitely could be. I'm just not sure if this is sane from a design perspective.

On one hand I agree, automating things would help the test writer. The problem comes when you want to have control over these things when analyzing issues identified by the test. I think we should follow the same workflow as if the network stack is configured with ip route. Another problem is that you as a test writer need to know that something is done automatically behind the scenes. Being explicit is imo better.

Reply:
> is a for loop really needed here, its the same amount of lines to do:
> 
> ```python
> host.eth0.up()
> host.br0.up()
> ```

Good point. I can change that.

Reply:
Fixed.

Reply:
> I think this should even be done automatically behind the scenes... Is it possible to get all created devices and call the `.up()` method on them in a single place? I don't think it's possible now, but it definitely could be. I'm just not sure if this is sane from a design perspective.

This is specifically because you want to be as explicit about what type of configuration you want to do for this specific test. Setting a device up/down during configuration is a very important step and as a test writer/user you want to know exactly when it happens. So the choice of doing it like this instead of behind the scenes was very much intentional.

At least for now I'd keep these as is, at some point we may want to take a closer look to see if it makes sense to refactor the test set implementations a bit more to improve some very obvious and common patterns but I don't think this is the right time for that.